LilitYeghiazaryan/BST

```diff
@@ -0,0 +1,680 @@
+    {
+
+            
+        return true;
+    }
+
+    void inorderRec() const
+    {
+        inorderRec(root);
+    }
+
+    // TODO
+    void preorderRec() const
+    {
+        preorderRec(root);
+    }
+
+    // TODO
+    void postorderRec() const
+    {
+        postorderRec(root);
+    }
+
+    // start here before inorder and postorder iterative versions
+    void preorder() const
+    {
+        Node* curr = root;
+        std::stack<Node*> stack1, stack2;
+        stack1.push(curr);
+
+        while (!stack1.empty())
+        {
+            curr = stack1.top();
+            stack1.pop();
+            std::cout << curr->data << ", ";
+
+            if (curr->right)
+            {
+                stack1.push(curr->right);
+            }
+            if (curr->left)
+            {
+                stack1.push(curr->left);
+            }
+
+        }
+        std::cout << std::endl;
+    }
+
+    // TODO
+    void inorder() const
+    {
+        Node* curr = root;
+        std::stack<Node*> stack1, stack2;
+
+
+        while (curr || !stack1.empty())
+        {
+            while (curr)
+            {
+                stack1.push(curr);
+                curr = curr->right;
+            }
+
+            curr = stack1.top();
+            stack1.pop();
+            stack2.push(curr);
+
+            curr = curr->left;
+        }
+
+
+        while (!stack2.empty())
+        {
+            std::cout << stack2.top()->data << ", ";
+            stack2.pop();
+        }
+        std::cout << std::endl;
+
+    }
+
+    // TODO
+    void postorder() const
+    {
+        Node* curr = root;
+        std::stack<Node*> stack1, stack2;
+        stack1.push(curr);
+
+        while (!stack1.empty())
+        {
+            curr = stack1.top();
+            stack1.pop();
+            if (curr->left)
+            {
+                stack1.push(curr->left);
+            }
+            if (curr->right)
+            {
+                stack1.push(curr->right);
+            }
+            stack2.push(curr);
+        }
+
+        while (!stack2.empty())
+        {
+            std::cout << stack2.top()->data << ", ";
+            stack2.pop();
+        }
+        std::cout << std::endl;
+
+    }
+
+    // TODO - using queue
+    int width() const
+    {
+        int max, temp_max, size;
+        Node* curr = root;
+        std::queue<Node*> queue;
+        queue.push(root);
+        max = root ? 1 : 0;
+
+        while (!queue.empty())
+        {
+            temp_max = 0;
+            size = queue.size();
+            while (size--)
+            {
+                curr = queue.front();
+                queue.pop();
+
+                if (curr->left)
+                {
+                    queue.push(curr->left);
+                    temp_max++;
+                }
+                if (curr->right)
+                {
+                    queue.push(curr->right);
+                    temp_max++;
+                }
+                if (temp_max > max)
```

Comment:
Why do you need to keep track of `temp_max` at all? `queue` size, when read at the right time, will give you the width of the current layer.